README.md
# H2O Wave POC

## Installation
Official guide: [h20-wave installation](https://wave.h2o.ai/docs/installation)

## Prepare environment

```shell
python3 -m venv .venv
pip install --upgrade pip
pip install -r requirements.txt
```

## Install wave server

*Ex: wave server 0.25.2 for Mac M1*
```shell
curl https://h2o-wave.s3.amazonaws.com/releases/wave-0.25.2-darwin-arm64.tar.gz --output wave-0.25.2-darwin-arm64.tar.gz
tar zxvf wave-0.25.2-darwin-arm64.tar.gz
ln -s wave-0.25.2-darwin-arm64 wave-server

```

## [Optional] Enable HTTP over TLS
Official guide: [Security](https://wave.h2o.ai/docs/security)

*Create TLS cert and key files*

```shell
cd wave-server
mkdir certs

openssl req \
   -newkey rsa:2048 -nodes -keyout domain.key \
   -x509 -days 365 -out domain.crt
```

*Export env variable for TLS*
```shell
export H2O_WAVE_TLS_CERT_FILE="./certs/domain.crt"
export H2O_WAVE_TLS_KEY_FILE="./certs/domain.key"
export H2O_WAVE_ADDRESS="https://127.0.0.1:10101"
```


## Run wave server
```shell
cd wave-server
./waved
```
> None
> If `H2O_WAVE_TLS_CERT_FILE` & `H2O_WAVE_TLS_KEY_FILE` are set, `wave` server will start over TLS.

## Run wave apps
> ! Must be located at root directory `h2o-poc/`
> 
```shell
wave run my_app.apps
```

CHANGELOG.md
# [0.0.2] - 2023-03-17

## doc - Update README

- Adding run wave apps

---
# [0.0.1] - 2023-03-17

## init - H2O Wave apps POC

- Init with python code
- README

---

my_app/apps.py
from h2o_wave import main, app, Q, ui, handle_on  # noqa

from my_app.project import pg_mc
from my_app.page2 import page2  # noqa
from my_app.page3 import page3  # noqa
from my_app.page4 import page4  # noqa


def zone_content():
    return ui.zone('content', zones=[
        # Specify various zones and use the one that is currently
        # needed. Empty zones are ignored.
        ui.zone('horizontal-1', direction=ui.ZoneDirection.ROW),
        ui.zone('horizontal-2', direction=ui.ZoneDirection.ROW),
        ui.zone('vertical-1'),
        ui.zone(
            'grid',
            direction=ui.ZoneDirection.ROW,
            wrap='stretch',
            justify='center'
        )
    ])


custom_theme_purple_dark = ui.theme(
    name='custom-theme-purple-dark',
    primary='#02111d',
    text='#483D8B',
    card='#E6E6FA',
    page='#483d8b'
)

custom_theme_purple_light = ui.theme(
    name='custom-theme-purple-light',
    primary='#1a1a22',
    text='#02111d',
    card='#E6E8FA',
    page='#C9CBE1'
)

desjardins_theme_green_dark = ui.theme(
    name='desjardins-theme-green-dark',
    primary='#065B37',
    text='#292B2F',
    card='#FFFFFF',
    page='#01874E'
)

desjardins_theme_green_dark_pro = ui.theme(
    name='desjardins-theme-green-dark-pro',
    primary='#292B2F',
    text='#065B37',
    card='#FFFBFF',
    page='#F2F2F2'
)

desjardins_theme_green_light = ui.theme(
    name='desjardins-theme-green-light',
    primary='#018654',
    text='#292B2F',
    card='#FFFFFF',
    page='#F2F2F2'
)
imgages = {
    "desjardins": [
        'https://www.desjardins.com/content/dam/images/logos/commun/logo-desjardins-fr.svg',
        'https://www.desjardins.com/ressources/images/d05-logo-ere-numerique2018.svg?resVer=1520861863000'
    ]
}


async def init(q: Q) -> None:
    q.page['meta'] = ui.meta_card(
        box='',
        themes=[
            custom_theme_purple_dark,
            custom_theme_purple_light,
            desjardins_theme_green_dark,
            desjardins_theme_green_light,
            desjardins_theme_green_dark_pro
        ],
        # theme='winter-is-coming',
        theme='desjardins-theme-green-dark-pro',
        layouts=[
            ui.layout(
                breakpoint='xs',
                min_height='100vh',
                zones=[
                    ui.zone(
                        'main', size='1', direction=ui.ZoneDirection.ROW,
                        zones=[
                            ui.zone('sidebar', size='250px'),
                            ui.zone(
                                'body',
                                zones=[
                                    ui.zone('header'),
                                    zone_content()
                                ]
                            ),
                        ]
                    )
                ]
            )
        ])

    q.page['sidebar'] = ui.nav_card(
        box='sidebar', color='primary', title='My App',
        subtitle="Let's conquer the world!",
        value=f'#{q.args["#"]}' if q.args['#'] else '#pg-mc',
        image=imgages['desjardins'][1],
        items=[
            ui.nav_group('Menu', items=[
                ui.nav_item(name='#pg-mc', label='Home'),
                ui.nav_item(name='#page2', label='Charts'),
                ui.nav_item(name='#page3', label='Grid'),
                ui.nav_item(name='#page4', label='Form'),
            ]),
        ])

    q.page['header'] = ui.header_card(
        box='header', title='', subtitle='',
        secondary_items=[ui.textbox(
            name='search', icon='Search',
            width='400px', placeholder='Search...')],
        items=[
            ui.persona(
                title='John Doe', subtitle='Developer', size='xs',
                image='https://images.pexels.com/photos/220453/'
                      'pexels-photo-220453.jpeg?auto=compress&h=750&w=1260'),
        ]
    )

    # If no active hash present, render page1.
    if q.args['#'] is None:
        await pg_mc(q)


@app('/')
async def serve(q: Q):
    # Run only once per client connection.
    if not q.client.initialized:
        q.client.cards = set()
        await init(q)
        q.client.initialized = True

    # Handle routing.
    await handle_on(q)
    await q.page.save()


my_app/page2.py
from h2o_wave import main, Q, ui, on, data  # noqa

from my_app.utils import clear_cards, add_card


@on('#page2')
async def page2(q: Q):
    # When routing, drop all the cards except of the main ones
    # (header, sidebar, meta).
    clear_cards(q)

    add_card(q, 'chart1', ui.plot_card(
        box='horizontal-1',
        title='Chart 1',
        data=data('category country product price', 10, rows=[
            ('G1', 'USA', 'P1', 124),
            ('G1', 'China', 'P2', 580),
            ('G1', 'USA', 'P3', 528),
            ('G1', 'China', 'P1', 361),
            ('G1', 'USA', 'P2', 228),
            ('G2', 'China', 'P3', 418),
            ('G2', 'USA', 'P1', 824),
            ('G2', 'China', 'P2', 539),
            ('G2', 'USA', 'P3', 712),
            ('G2', 'USA', 'P1', 213),
        ]),
        plot=ui.plot(
            [ui.mark(
                type='interval',
                x='=product',
                y='=price',
                color='=country',
                stack='auto',
                dodge='=category',
                y_min=0
            )]
        )
    ))
    add_card(q, 'chart2', ui.plot_card(
        box='horizontal-1',
        title='Chart 2',
        data=data('date price', 10, rows=[
            ('2020-03-20', 124),
            ('2020-05-18', 580),
            ('2020-08-24', 528),
            ('2020-02-12', 361),
            ('2020-03-11', 228),
            ('2020-09-26', 418),
            ('2020-11-12', 824),
            ('2020-12-21', 539),
            ('2020-03-18', 712),
            ('2020-07-11', 213),
        ]),
        plot=ui.plot(
            [ui.mark(
                type='line',
                x_scale='time',
                x='=date',
                y='=price',
                y_min=0
            )]
        )
    ))
    add_card(q, 'table', ui.form_card(box='vertical-1', items=[ui.table(
        name='table',
        downloadable=True,
        resettable=True,
        groupable=True,
        columns=[
            ui.table_column(name='text', label='Process', searchable=True),
            ui.table_column(
                name='tag',
                label='Status',
                filterable=True,
                cell_type=ui.tag_table_cell_type(
                    name='tags',
                    tags=[
                        ui.tag(label='FAIL', color='$red'),
                        ui.tag(
                            label='DONE',
                            color='#D2E3F8',
                            label_color='#053975'
                        ),
                        ui.tag(label='SUCCESS', color='$mint'),
                    ]
                )
            )
        ],
        rows=[
            ui.table_row(name='row1', cells=['Process 1', 'FAIL']),
            ui.table_row(name='row2', cells=['Process 2', 'SUCCESS,DONE']),
            ui.table_row(name='row3', cells=['Process 3', 'DONE']),
            ui.table_row(name='row4', cells=['Process 4', 'FAIL']),
            ui.table_row(name='row5', cells=['Process 5', 'SUCCESS,DONE']),
            ui.table_row(name='row6', cells=['Process 6', 'DONE']),
        ])
    ]))


my_app/page4.py
from h2o_wave import main, Q, ui, on  # noqa

from my_app.utils import clear_cards, add_card


@on('#page4')
async def page4(q: Q):
    # When routing, drop all the cards except of the main ones
    # (header, sidebar, meta).
    # Since this page is interactive, we want to update its card
    # instead of recreating it every time, so ignore 'form' card on drop.
    clear_cards(q, ['form'])

    if q.args.step1:
        # Just update the existing card, do not recreate.
        q.page['form'].items = [
            ui.stepper(name='stepper', items=[
                ui.step(label='Step 1'),
                ui.step(label='Step 2'),
                ui.step(label='Step 3'),
            ]),
            ui.textbox(name='textbox2', label='Textbox 1'),
            ui.buttons(justify='end', items=[
                ui.button(name='step2', label='Next', primary=True),
            ])
        ]
    elif q.args.step2:
        # Just update the existing card, do not recreate.
        q.page['form'].items = [
            ui.stepper(name='stepper', items=[
                ui.step(label='Step 1', done=True),
                ui.step(label='Step 2'),
                ui.step(label='Step 3'),
            ]),
            ui.textbox(name='textbox2', label='Textbox 2'),
            ui.buttons(justify='end', items=[
                ui.button(name='step1', label='Cancel'),
                ui.button(name='step3', label='Next', primary=True),
            ])
        ]
    elif q.args.step3:
        # Just update the existing card, do not recreate.
        q.page['form'].items = [
            ui.stepper(name='stepper', items=[
                ui.step(label='Step 1', done=True),
                ui.step(label='Step 2', done=True),
                ui.step(label='Step 3'),
            ]),
            ui.textbox(name='textbox3', label='Textbox 3'),
            ui.buttons(justify='end', items=[
                ui.button(name='step2', label='Cancel'),
                ui.button(name='submit', label='Next', primary=True),
            ])
        ]
    else:
        # If first time on this page, create the card.
        add_card(q, 'form', ui.form_card(box='vertical-1', items=[
            ui.stepper(name='stepper', items=[
                ui.step(label='Step 1'),
                ui.step(label='Step 2'),
                ui.step(label='Step 3'),
            ]),
            ui.textbox(name='textbox1', label='Textbox 1'),
            ui.buttons(justify='end', items=[
                ui.button(name='step2', label='Next', primary=True),
            ]),
        ]))


my_app/project.py
from h2o_wave import main, Q, ui, on  # noqa

from my_app.utils import clear_cards, add_card


@on('#pg-mc')
async def pg_mc(q: Q):
    # When routing, drop all the cards except of the main ones
    # (header, sidebar, meta).
    clear_cards(q)

    add_card(
        q, 'project',
        ui.tall_info_card(
            box='horizontal-1',
            name='#project',
            title='Project',
            caption='',
            icon='People'
        )
    )

    add_card(
        q, 'Component',
        ui.tall_info_card(
            box='horizontal-2',
            name='#component',
            title='Component',
            caption='',
            icon='ViewInAR'
        )
    )

    add_card(
        q, 'MetaComponent',
        ui.tall_info_card(
            box='horizontal-2',
            name='#meta_component',
            title='MetaComponent',
            caption='',
            icon='ViewAll2'
        )
    )

    add_card(
        q, 'article',
        ui.tall_article_preview_card(
            box=ui.box('vertical-1', height='600px'),
            title='How does magic work',
            image='https://images.pexels.com/photos/624015/pexels-photo-624015.jpeg?auto=compress&cs=tinysrgb&dpr=1&h=750&w=1260',
        content='''
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum ac sodales felis. Duis orci enim, iaculis at augue vel, mattis imperdiet ligula. Sed a placerat lacus, vitae viverra ante. Duis laoreet purus sit amet orci lacinia, non facilisis ipsum venenatis. Duis bibendum malesuada urna. Praesent vehicula tempor volutpat. In sem augue, blandit a tempus sit amet, tristique vehicula nisl. Duis molestie vel nisl a blandit. Nunc mollis ullamcorper elementum.
Donec in erat augue. Nullam mollis ligula nec massa semper, laoreet pellentesque nulla ullamcorper. In ante ex, tristique et mollis id, facilisis non metus. Aliquam neque eros, semper id finibus eu, pellentesque ac magna. Aliquam convallis eros ut erat mollis, sit amet scelerisque ex pretium. Nulla sodales lacus a tellus molestie blandit. Praesent molestie elit viverra, congue purus vel, cursus sem. Donec malesuada libero ut nulla bibendum, in condimentum massa pretium. Aliquam erat volutpat. Interdum et malesuada fames ac ante ipsum primis in faucibus. Integer vel tincidunt purus, congue suscipit neque. Fusce eget lacus nibh. Sed vestibulum neque id erat accumsan, a faucibus leo malesuada. Curabitur varius ligula a velit aliquet tincidunt. Donec vehicula ligula sit amet nunc tempus, non fermentum odio rhoncus.
Vestibulum condimentum consectetur aliquet. Phasellus mollis at nulla vel blandit. Praesent at ligula nulla. Curabitur enim tellus, congue id tempor at, malesuada sed augue. Nulla in justo in libero condimentum euismod. Integer aliquet, velit id convallis maximus, nisl dui porta velit, et pellentesque ligula lorem non nunc. Sed tincidunt purus non elit ultrices egestas quis eu mauris. Sed molestie vulputate enim, a vehicula nibh pulvinar sit amet. Nullam auctor sapien est, et aliquet dui congue ornare. Donec pulvinar scelerisque justo, nec scelerisque velit maximus eget. Ut ac lectus velit. Pellentesque bibendum ex sit amet cursus commodo. Fusce congue metus at elementum ultricies. Suspendisse non rhoncus risus. In hac habitasse platea dictumst.
        '''
        )
    )

#


@on('#project')
async def home(q: Q):
    clear_cards(q)

    add_card(
        q=q,
        name='result_form',
        card=ui.markdown_card(
            box='horizontal-1',
            title='Result form',
            content='Data here'
        )
    )

    add_card(
        q=q,
        name='form',
        card=ui.form_card(
            box='horizontal-2',
            items=[
                ui.textbox(name='project_name', label='Project name', ),
                ui.textbox(name='email_group', label='Email group',
                           placeholder='email_group@desjardins.com'),
                ui.button(name='create_project', label='Create', primary=True),
            ]
        )
    )

    if q.args.create_project:
        # Update existing card content.
        q.page['result_form'].content = \
            f"{q.args.email_group} - {q.args.project_name}"
